samples/auto_layout.lua

```diff
@@ -0,0 +1,349 @@
+function plugindef()
+    finaleplugin.RequireDocument = false
+    finaleplugin.MinJWLuaVersion = 0.71
+    return "0--auto resize width test"
+end
+
+local utils = require('library.utils')
+local mixin = require('library.mixin')
+local localization = require('library.localization')
+
+--
+-- For scripts in the `src` directory, each localization should be separately stored in the
+-- `localization` subdirectory. See the comments in `library/localization.lua` for more details.
+-- The localization tables are included here in this sample to keep the sample self-contained.
+--
+
+--
+-- This table was auto-generated with `utilities/localization_tool.lua`
+-- Then it was edited to include only the strings that need to be localized.
+--
+localization.en = -- this is en_GB due to spelling of "Localisation"
+{
+    ["Action Button"] = "Action Button",
+    ["Choices"] = "Choices",
+    ["Close"] = "Close",
+    ["First Option"] = "First Option",
+    ["Fourth Option"] = "Fourth Option",
+    ["Left Checkbox Option 1"] = "Left Checkbox Option 1",
+    ["Left Checkbox Option 2"] = "Left Checkbox Option 2",
+    ["Menu"] = "Menu",
+    ["Right Three-State Option"] = "Right Three-State Option",
+    ["Second Option"] = "Second Option",
+    ["Short "] = "Short ",
+    ["Test Autolayout With Localisation"] = "Test Autolayout With Localisation",
+    ["Third Option"] = "Third Option",
+    ["This is long menu text "] = "This is long menu text ",
+    ["This is long text choice "] = "This is long text choice ",
+    ["This is longer option text "] = "This is longer option text ",
```

Comment:
I still submit that encouraging authors to use full language strings as keys is inviting a world of pain...

Reply:
>is inviting a world of pain...

Perhaps. Full language strings are encouraged (not required) in macOS localization. The advantages are

- the code is more readable
- the code can be written and tested before any considerations of localization, beyond using the `*Localized` methods.
- the text injected into the control is guaranteed to be human readable in some language.
- the solution is agnostic as to which language the programmer chooses for the base language.

That said, I have encouraged using non-human-readable keys in the `plugindef`. I would (in particular) hope that if we ever have to localize `finaleplugin.Notes` it will be with some key other than the full text. There is nothing requiring full language keys, but then there has to be a fallback language. I would rather leave that bridge uncrossed at the moment.

Reply:
Why do these strings include the trailing space?

Reply:
> Full language strings are encouraged (not required) in macOS localization

I find that very surprising. If you look at various open source code bases (or even docs of some proprietary software), you'll be hard pressed to find i18n approached like that. For example: https://github.com/lokalise/i18n-ally/tree/main/locales which also utilises a kind of namespacing for grouping its strings. Or you can try any forum software, eg https://xenforo.com/docs/xf2/phrases/ or https://github.com/SimpleMachines/SMF/tree/release-3.0/Languages I'd encourage you to check out others too.

&nbsp;
> * the code is more readable

Not necessarily, as it can also make it more verbose. If I'm flicking through something, I'd much rather see this
```lua
finenv.UI():AlertError(loc.localize("error_msg_transposition"), loc.localize("error_title"))
```
than this
```lua
finenv.UI():AlertError(
            loc.localize(
                "Finale is unable to represent some of the transposed pitches. These pitches were left unchanged."),
            loc.localize("Transposition Error")
        )
```
I also find it gets in the way of separation of concerns. But more in the next dot point.

&nbsp;
> * the code can be written and tested before any considerations of localization, beyond using the `*Localized` methods.

Generally, a piece of code is either written with localisation in mind or it isn't. The current approach attempts to straddle the line so you get code that doesn't have to be localised, but it could be and I don't think it works. It relies too much on the magic of fallbacks rather than having well written and well structured code from the start. In the process you also get language string keys that don't look like keys, and for which you don't get the benefit of autocomplete, and which are also harder to edit.

For example, let's say I want to change the wording of an error message. If the key is `user_error` then I just have to change the value of `loc.user_error` (and generate a new set of translations if needed). But if the key is `"Fatal user error, please step away from the keyboard as it will self destruct in 3 ... 2 ... 1 ..."` then I need to change the key as well, or else every time I see the message in the code, I'm going to think "hold on, I changed that". So now instead of just editing the string, I have to highlight the text and use find and replace (and hope no errors were introduced). If for whatever reason it was a minor edit (like capitalisation or punctuation) and regenerating the translations wasn't necessary, I'd have to do that anyway or at least remember to replace the keys in _every_ language file.

There's also the matter using the keys, especially if you need them in multiple places. If I'm typing and I want to use a language string, I have to go back and copy and paste the whole thing. Typing from memory isn't even reliable since all keys in Lua are case-sensitive. That's not easier than typing `user_error` or even just getting part way through and hitting enter is easier.

Code written with i18n in mind will always look different to code that isn't: the content of the language string doesn't matter any more, what matters is what it is, which can also include when/where it is used (eg `menu_item_1`, `menu_item_2`, `exit`, `unknown_error`, etc). And that's how keys are named, according to what it is, which isn't necessarily what it says. When you're reading the code, you can focus on the code itself. When you're using the strings, the keys are easy to type and remember (or autocomplete). Then if you want to edit the copy, just look at the language file and focus on that.

&nbsp;
> * the text injected into the control is guaranteed to be human readable in some language.
> 
> * the solution is agnostic as to which language the programmer chooses for the base language.

Both of these could be better handled by having a default language (eg `en`) for which a language file is required to exist (if it doesn't, throw an error). If the author wants to change this, they could do so by calling a function (eg `localization.set_default_locale('fr')` ).

&nbsp;
> I would rather leave that bridge uncrossed at the moment.

I18n is something that's much easier to handle if you spend a little time to plan it out and get it right from the start,  before it becomes unwieldy with vast amounts of language files and keys/strings.